### PR TITLE
Support other prngs

### PR DESCRIPTION
**Pull request by rolyatmax**
This PR adds support for optionally passing a custom `random` function to be used instead of `Math.random`.

See: https://github.com/d3/d3-random/issues/6

@mbostock Let me know if this is going down the right path, or if you'd prefer a `randomSource` function as you mentioned in [this comment](https://github.com/d3/d3-random/issues/6#issuecomment-176984568).

Also, I still need to add tests for this!

### CODE REVIEW

**Comment by rolyatmax** [comment]
Also not sure if it's a bit tedious to call out the optional `random` argument for each function in the docs, or if that should be mentioned just once somewhere else. Happy to clean those up depending on how others feel about it!

**Comment by mbostock** [comment]
How do you plan on using this new functionality? For a seeded random number generator? (Part of me wonders whether if this is just for seeded random if a more restrictive API would be better, such as one that takes a seed rather than a function. But taking a function is probably fine; it is more flexible in exchange for being easier to misuse.)

This approach does feel a bit invasive, and I am always wary of adding extra optional arguments. I would like to see the d3.randomSource approach for comparison, returning an object that exposes the random number generators (using the associated source of randomness, a.k.a., the given base uniform random number generator in [0,1]).

**Comment by rolyatmax** [comment]
> How do you plan on using this new functionality? For a seeded random number generator?

Yup! I often create visualizations or generative art that I would like to reproduce exactly and having a seeded prng becomes necessary. Two examples:

1. Often when I'm prototyping, I'll take a variable and adjust its values (using a tool like `dat-gui`) to see how it effects the visualization. In order for this to work, I hold all other values constant. I can do this with a random number generator by reseeding it with the same value at the start of every render.

2. If I'm working with a large dataset or a large image output size, I'll work on a scaled down version of a visualization while prototyping. This speeds up the development feedback loop by reducing the amount of time I have to wait between renders. When it comes time to output the final, full-scale image, however, I often want it to use the same random values so that it looks the same (or almost the same) as what I was seeing in the prototype.

> Part of me wonders whether if this is just for seeded random if a more restrictive API would be better, such as one that takes a seed rather than a function.

I don't feel super strongly about this, but I can think of two_possible_ advantages of passing a function instead of a seed:

1. It _might_ make testing a bit easier/cleaner - both for this library and for code that uses this library. Using a "random" function which returns a known value (e.g. `() => 0.5`) would help ensure that you're testing just the non-random-number-generating logic in your library. This might not be a huge win for `d3-random`'s tests (they seem to work well as is) but could make testing code which relies on `d3-random` a bit easier.

2. It allows users to compose multiple random functions:
```js
const normal = d3.randomNormal(-1, 1)
const combinedNormal = d3.randomNormal(1, 1, normal)
```
This _seems_ to be useful, but I have to admit I haven't come across a practical need for this before. (Though, now I'm pretty sure I'm gonna start using this in my generative art!)

Let me know what you think about all this. In the meantime, I'll give the `d3.randomSource` approach a try!

**Comment by mbostock** [comment]
I think d3.randomSource(*function*) is probably the way to go. For one thing it means we don’t need a dependency on a specific implementation of a seeded pseudorandom number generator.

However I do think it’s important to document the constraints of this function—that it is expected to return a random float in the range [0,1) (like Math.random). So composing random functions as you suggest would not be supported in general (though may still be *possible* in some cases), as in your example the `normal` does not comply with this interface because it can return values outside the range [0,1).

**Comment by rolyatmax** [comment]
> as in your example the normal may return values outside the range [0,1)
Ah yes! Good catch - I think you're absolutely right re: the [0, 1) range.

I'll find some time over the next few days to work on the `d3.randomSource` approach.

Thanks for the quick feedback!

**Comment by rolyatmax** [comment]
I took a pass at implementing the `randomSource` approach. A few things to note:

1. To make the random number generator used in each function changeable, I've wrapped each in a function that takes a random number generator and returns the d3-random function bound to the given random number generator. This also feels invasive to me, though I'm not sure how else we'd swap out `Math.random` in the functions that doesn't involve messing with the global `Math.random`.

2. You'll notice I'm not leveraging the logic in `src/source.js` when exporting the default versions of the module's functions in `index.js`. This is because I'm trying to preserve some of the tree-shaking benefits of ES6 modules. As it is stands, `randomSource` returns an object with all the random functions attached to it, so tree-shaking wouldn't remove the unused functions from the resulting bundle. I _believe_ exporting each function's `Math.random`-bound version in `index.js` preserves those tree-shaking benefits. (I'm not an expert on how rollup, et al implement tree-shaking, so I could be mistaken here.)

3. Taking my cues from the simplicity of the codebase, I chose not to include checks to make sure the `randomFn` arg is a function. I don't feel strongly about this one way or the other, though.

I think all things considered, I might prefer the optional arg approach just slightly to the `randomSource` approach. Neither dramatically alters the existing API, but the optional arg approach seems to be a little less invasive to the code itself (mostly just adds the optional arg as opposed to wrapping each function). Perhaps I'm missing a cleaner way to do this, though.

Let me know what you think about all this! :-)

**Comment by mbostock** [comment]
There’s a third pattern here we can adopt from d3-interpolate. It looks like this:

```js
var value = d3.randomUniform.source(myRandom)(min, max);
```

Or with the default source, Math.random, as today:

```js
var value = d3.randomUniform(min, max);
```

The implementation looks like this:

```js
export default (function randomUniform(source) {

  function random(min, max) {
    min = min == null ? 0 : +min;
    max = max == null ? 1 : +max;
    if (arguments.length === 1) max = min, min = 0;
    else max -= min;
    return function() {
      return source() * max + min;
    };
  }

  random.source = randomUniform;

  return random;
})(Math.random);
```

That seems cleaner than the optional argument from an interface perspective and retains tree-shaking.

**Comment by rolyatmax** [comment]
I'm liking this approach. I'll take a swing at it tonight or tomorrow. Thanks for the feedback!

**Comment by mbostock** [comment]
Looks great. Mind adding a few tests?

**Comment by rolyatmax** [comment]
Sure thing! Will add some tests and update docs. Currently investigating a test failure with Bates/IrwanHall that shows up occasionally.

**Comment by rolyatmax** [comment]
@mbostock thoughts on turning on `semi` in the `.eslintrc`?

**Comment by rolyatmax** [comment]
So one thing I discovered: this implementation does seem to break runtime `Math.random` monkey-patching since the `Math.random` function is cached within the `createRandomFunction`'s scope. So this change would _require_ users to start using `source(fn)` for custom random number generators.

We could also wait until the random function is called before falling back to `Math.random`. But it's kinda ugly:

```js
export default (function sourceRandomUniform(source) {
  function randomUniform(min, max) {
    min = min == null ? 0 : +min;
    max = max == null ? 1 : +max;
    if (arguments.length === 1) max = min, min = 0;
    else max -= min;
    return function() {
      source = source == null ? Math.random : source;
      return source() * max + min;
    };
  }

  randomUniform.source = sourceRandomUniform;

  return randomUniform;
})();
```

**Comment by rolyatmax** [comment]
Also, got tests passing by replacing the `Math.random` monkey-patching with the `.source(fn)` method in the tests, but feel like we're now neglecting to test the default behavior. Perhaps it's sufficient to just test that the random functions return numbers even when the source isn't set?

**Comment by mbostock** [comment]
The simple fix to allow monkey-patching again is to use this in place of Math.random:

```js
function defaultSource() {
  return Math.random();
}
```

I dunno if it’s required that we do this, but it doesn’t really feel like it would hurt to do, either.

**Comment by rolyatmax** [comment]
Sounds good to me. Do you prefer one of these over the other?:
```js
import defaultSource from "./defaultSource";
export default (function sourceRandomUniform(source) {
  // ...
})(defaultSource);
```

```js
export default (function sourceRandomUniform(source) {
  // ...
})(function() {
  return Math.random();
});
```

**Comment by mbostock** [comment]
The former (sharing the definition).

File: src/exponential.js [hunk 1]
```diff
@@ -1,5 +1,12 @@
-export default function(lambda) {
-  return function() {
-    return -Math.log(1 - Math.random()) / lambda;
-  };
-}
+import defaultSource from "./defaultSource";
+export default (function sourceRandomExponential(source) {
```
Comment (mbostock) [inline src/exponential.js]: Blank line above this line?

File: src/irwinHall.js [hunk 2]
```diff
@@ -1,6 +1,13 @@
-export default function(n) {
-  return function() {
-    for (var sum = 0, i = 0; i < n; ++i) sum += Math.random();
-    return sum;
-  };
-}
+import defaultSource from "./defaultSource";
+export default (function sourceRandomIrwinHall(source) {
```
Comment (mbostock) [inline src/irwinHall.js]: Same here.

File: src/normal.js [hunk 3]
```diff
@@ -1,20 +1,27 @@
-export default function(mu, sigma) {
-  var x, r;
-  mu = mu == null ? 0 : +mu;
-  sigma = sigma == null ? 1 : +sigma;
-  return function() {
-    var y;
+import defaultSource from "./defaultSource";
+export default (function sourceRandomNormal(source) {
```
Comment (mbostock) [inline src/normal.js]: And here.

File: src/uniform.js [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
-  min = min == null ? 0 : +min;
-  max = max == null ? 1 : +max;
-  if (arguments.length === 1) max = min, min = 0;
-  else max -= min;
-  return function() {
-    return Math.random() * max + min;
-  };
-}
+import defaultSource from "./defaultSource";
+export default (function sourceRandomUniform(source) {
```
Comment (mbostock) [inline src/uniform.js]: And here.

**Comment by mbostock** [comment]
Sure, a test for monkey-patching would be nice, but it should make clear that it is not a requirement of the API and we might remove it in the future without a major version change.

**Comment by mbostock** [comment]
Thanks for the contribution! Released as 1.1.0.

**Comment by rolyatmax** [comment]
No problem! Happy to help!

Was working on some tests to ensure Math.random is called by default. (Am still sorting out some issues with monkey-patching within tape.) Happy to keep working on those if you'd like.

**Pull request merged**